stdlib/public/Platform/Windows/SwiftWinDbg/SwiftWinDbg.cpp

```diff
@@ -0,0 +1,295 @@
+//DECLARE_API(readarray) {
+//	const ULONG64 CountOffset = 16, CapacityOffset = 24, BufferOffset = 32;
+//
+//	// FIXME: Can I lookup the stride in the .pdb?
+//	PSTR TypeStr = "Swift::Int64";
+//
+//	ULONG_PTR ArrayPointerAddress = GetExpression(args);
+//	if (ArrayPointerAddress == 0) {
+//		dprintf("Unable to find address of %s\n", args);
+//		return;
+//	}
+//	PCSTR Name = args;
+//	//dprintf("%s = %p\n", Name, ArrayPointerAddress);
+//
+//	ULONG Stride = GetTypeSize(TypeStr);
+//	if (Stride == 0) {
+//		dprintf("Unable to get size of %s\n", TypeStr);
+//		return;
+//	}
+//
+//	ULONG_PTR ArrayPointer;
+//	if (!ReadMemory(ArrayPointerAddress, &ArrayPointer, sizeof(ArrayPointer),
+//                  NULL)) {
+//		dprintf("Unable to read array at %p\n", ArrayPointerAddress);
+//		return;
+//	}
+//	//dprintf("*%s = %p\n", Name, ArrayPointer);
+//
+//	ULONG Count;
+//	if (!ReadMemory(ArrayPointer + CountOffset, &Count, sizeof(Count), NULL)) {
+//		dprintf("Unable to read array count at %p\n", ArrayPointer + CountOffset);
+//		return;
+//	}
+//	//dprintf("Swift Array at %p has %d elements\n", ArrayPointer, Count);
+//
+//	PUCHAR ArrayBuffer = (PUCHAR)malloc(Count * Stride);
+//	if (!ReadMemory(ArrayPointer + BufferOffset, ArrayBuffer, Count * Stride,
+//                  NULL)) {
+//		dprintf("Unable to read array buffer at %p\n",
+//            ArrayPointer + BufferOffset);
+//		return;
+//	}
+//
+//	// ([TYPE]) Name = COUNT values {
+//	//		[0] = BYTES
+//	//		[1] = BYTES...
+//	// }
+//	dprintf("([%s]) %s = %d values {", TypeStr, Name, Count);
+//	for (ULONG idx = 0; idx < Count; idx++) {
+//		dprintf("\n\t[%d] =", idx);
+//		for (ULONG byteOffset = 0; byteOffset < Stride; byteOffset++) {
+//			dprintf(" %02x", ArrayBuffer[Stride * idx + byteOffset]);
+//		}
+//	}
+//	dprintf("\n}\n");
+//	free(ArrayBuffer);
+//}
+
+DECLARE_API(readstring) {
+	ULONG_PTR StringGuts = GetExpression(args);
+	if (StringGuts == 0) {
+		dprintf("Could not find address of %s\n", args);
+		return;
+	}
+	PCSTR Name = args;
+	//dprintf("_guts = %p\n", StringGuts);
+
+	ULONG_PTR StringObject;
+	if (GetFieldValue(StringGuts, "Swift::_StringGuts", "_object",
+		StringObject)) {
+		dprintf("Unable to get _object field from Swift::_StringGuts at %p\n",
+			StringGuts);
+		return;
+	}
+	//dprintf("_object = %p\n", StringObject);
+
+	ULONG_PTR StringOtherBits;
+	if (GetFieldValue(StringGuts, "Swift::_StringGuts", "_otherBits",
+		StringOtherBits)) {
+		dprintf("Unable to get _otherBits field from Swift::_StringGuts at %p\n",
+			StringGuts);
+		return;
+	}
+	//dprintf("_otherBits = %p\n", StringOtherBits);
+
+	// See StringGuts_SummaryProvider() at
+	// swift-lldb/source/Plugins/Language/Swift/SwiftFormatters.cpp
+	bool isValue = StringObject & (1ULL << 63);
+	bool isCocoaOrSmall = StringObject & (1ULL << 62);
+	bool isOpaque = StringObject & (1ULL << 61);
+	bool isUTF16 = StringObject & (1ULL << 60);
+	ULONG_PTR PayloadAddress = StringObject & ((1ULL << 56) - 1);
```

Comment:
Is there any reason to not use standard types as in, e.g. `uint32_t` ?

Reply:
I guess I was following the coding styles of other WinDbg extensions. I'll reformat this file before I'm finished.